geomesa-hbase/geomesa-hbase-rpc/src/main/scala/org/locationtech/geomesa/hbase/rpc/coprocessor/GeoMesaHBaseCallBack.scala

```diff
@@ -8,16 +8,65 @@
 
 package org.locationtech.geomesa.hbase.rpc.coprocessor
 
-import java.util.concurrent.LinkedBlockingQueue
+import java.io.InterruptedIOException
+import java.util.Base64
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.{CancellationException, LinkedBlockingQueue}
 
-import com.google.protobuf.ByteString
+import com.google.protobuf.{ByteString, RpcCallback}
 import com.typesafe.scalalogging.LazyLogging
-import org.apache.hadoop.hbase.client.coprocessor.Batch.Callback
-import org.locationtech.geomesa.hbase.proto.GeoMesaProto.GeoMesaCoprocessorResponse
+import org.apache.hadoop.hbase.client.Scan
+import org.apache.hadoop.hbase.client.coprocessor.Batch.{Call, Callback}
+import org.apache.hadoop.hbase.protobuf.ProtobufUtil
+import org.locationtech.geomesa.hbase.proto.GeoMesaProto.{GeoMesaCoprocessorRequest, GeoMesaCoprocessorResponse, GeoMesaCoprocessorService}
+import org.locationtech.geomesa.hbase.rpc.coprocessor.GeoMesaCoprocessor.{GeoMesaHBaseRequestVersion, ScanOpt}
 import org.locationtech.geomesa.utils.index.ByteArrays
 
-class GeoMesaHBaseCallBack(result: LinkedBlockingQueue[ByteString]) extends Callback[GeoMesaCoprocessorResponse] with LazyLogging {
-  var lastRow: Array[Byte] = _
+/**
+ * Callback class for invoking htable services
+ *
+ * @param scan initial scan (warning: may be mutated)
+ * @param options coprocessor configuration options
+ * @param result queue for storing results
+ * @param closed indicates if the call has been closed and we should stop querying
+ */
+class GeoMesaHBaseCallBack(
+    scan: Scan,
+    options: Map[String, String],
+    result: LinkedBlockingQueue[ByteString],
+    closed: AtomicBoolean
+  ) extends Callback[GeoMesaCoprocessorResponse] with Iterator[Scan] with LazyLogging {
+
+  private var updated: Array[Byte] = _ // tracks our status per invocation to detect multiple region updates
+
+  val callable: Call[GeoMesaCoprocessorService, GeoMesaCoprocessorResponse] =
+    new Call[GeoMesaCoprocessorService, GeoMesaCoprocessorResponse]() {
+      override def call(instance: GeoMesaCoprocessorService): GeoMesaCoprocessorResponse = {
+        if (closed.get) { null } else {
+          val controller = new GeoMesaHBaseRpcController()
+          val callback = new RpcCallbackImpl()
+          // note: synchronous call
+          try { instance.getResult(controller, buildRequest, callback) } catch {
+            case _: InterruptedException | _: InterruptedIOException | _: CancellationException =>
+              logger.warn("Cancelling remote coprocessor call")
+              controller.startCancel()
+          }
+
+          if (controller.failed()) {
+            logger.error(s"Controller failed with error:\n${controller.errorText()}")
+          }
+
+          callback.get()
+        }
+      }
+    }
+
+  override def hasNext: Boolean = !closed.get() && scan.getStartRow != null
```

Comment:
*Blinks*  What's up with the 'scan.getStartRow != null' condition?!

Reply:
I set it `null` down on line 89 as an end condition: https://github.com/locationtech/geomesa/pull/2533/files/ae9b5f3182d75fb7192250de8a0ce3a10951e3d6#diff-4e356118c06df279e1a802c46b56d5deR89
I am fairly sure that we don't ever set it null in a normal scan, as we use an empty byte array instead.

Reply:
alternatively, to make it clearer, we could have a separate variable to track the same thing. i didn't go that route b/c the scan class didn't complain about it being null and it seemed redundant to have a separate var.

Reply:
Ah, ok!  That makes sense.  Would it be worth adding a comment about the fact that we are using 'null' as a sentinel?  And otherwise, we expect an empty byte array?

Reply:
added